src/data_generators/publisher_data_file.py

```diff
@@ -0,0 +1,198 @@
+from wfa_planning_evaluation_framework.data_generators.impression_generator import (
+    ImpressionGenerator,
+)
+from wfa_planning_evaluation_framework.data_generators.pricing_generator import (
+    PricingGenerator,
+)
+
+
+class PublisherDataFile:
+    """Real or simulated impression log data for a single publisher.
+
+    A PublisherDataFile represents a viewing log for a single publisher.
+    It can be thought of as a sequence of events of the form (id, spend),
+    where id is the (virtual) id of a user, and spend is the total spend
+    amount that would cause this impression to be shown.
+
+    A PublisherDataFile object can be constructed in one of several ways:
+      1. It can be constructed directly from a list of (id, spend) tuples.
+      2. It can be read from a flat file.
+      3. It can be generated randomly.
+
+    In the last case, e.g. random generation, the generation of the publisher
+    data file is divided into two steps, which are orthogonal to each other.
+    In the first step, the sequence of id's is generated, with multiplicities.
+    In the second step, spend information is associated to the id's.  Thus,
+    to generate a random PublisherDataFile, two objects must be given: an
+    ImpressionGenerator object and a PriceGenerator object.
+    """
+
+    def __init__(
+        self, impression_log_data: Iterable[Tuple[int, float]], name: str = None
+    ):
+        """Constructs a PublisherDataFile object from raw event data.
+
+        Internally, the data is sorted in increasing order by total spend.
+        To fetch the users reached for a given spend, the prefix of events
+        up to that spend are taken.  To fetch the users reached for an
+        impression buy of N impressions, the first N events in the sorted
+        list are taken.
+
+        Args:
+          impression_log_data: An iterable consisting of pairs (id, spend),
+            where id is a (virtual) user id, and spend is the total amount
+            that would need to be spent in order for this impression to be
+            shown.
+          name: If specified, a human-readable name that will be associated
+            to this publisher data file.  For example, it could be an encoding
+            of the parameters that were used to create this publisher data file,
+            such as "homog_n10000_lambda5".
+        """
+        self._data = deepcopy(impression_log_data)
+        self._data.sort(key=lambda x: x[1])  # sort by spend
+        self._name = name
+        self._spend = max([spend for (_, spend) in impression_log_data])
+        self._reach = len(set([id for (id, _) in impression_log_data]))
+
+    @property
+    def impressions(self):
+        """Total number of impressions represented by this PublisherDataFile."""
+        return len(self._data)
+
+    @property
+    def spend(self):
+        """Total spend represented by this PublisherDataFile."""
+        return self._spend
+
+    @property
+    def reach(self):
```

Comment:
In the reach_curve PR, we used the term "max_reach". It seems that would be a better name for it given this is the max_possible reach this publisher can have, but totally up to you, just wanted to point out an alternative.

Reply:
Done.